Replace when.js with native promises and a polyfill

tbroyer opened the pull request:
Tested non-regression in Chrome (51/Linux), Firefox (49/Dev Edition/Linux) and Edge (13).
Couldn't actually test that it works in IE11 (the only browser actually needing the polyfill) due to other errors.

Note: "bundles" the polyfill in the library for now; we may want to revisit this in the future.

Close #97 

r: @peppsac

peppsac commented:
Looks good.

A few places still reference `when.js`:
- packages.json
- LICENSE.md (and maybe you could add an entry from es6-promise here as well?)
- ManagerCommands comment (line 11)
- Ori.js
- ProjectiveTexturingMaterial.js
- KMZLoader

tbroyer committed:
Replace when.js with native promises and a polyfill

Does not fix concurrency bugs or subobtimal resource usage where code
has been changed, but add comments to identify those places.

tbroyer committed:
Remove Require.js from LICENSE.md

Require.js is no longer used since the (early) switch to webpack.

tbroyer committed:
Fix suboptimal KMZ cache in KML_Provider

The cache now stores promises, whether resolved or not. That way, when
the parseKML method is called a second time for the same KMZ, before
that KMZ's loading has resolved, it will simply return the cached promise
instead of loading the file again. Once resolved, the promise can still
be used directly, so it's kept in the cache (no need to replace it).

tbroyer committed:
Fix concurrency bug in ProjectiveTexturingMaterial.init()

In case init() is called a second time before the it has resolved,
it now returns the same (unresolved) promise instead of a resolved
promise of 'null' (because _shaderMat wouldn't have been initialized
yet). Once resolved, the promise can still be used directly, so it's
only a matter of creating one only once.

tbroyer committed:
Refactor some nested promises (aka "callback pyramid of doom")

tbroyer commented:
Totally removed `when.js`, and also refactored a few things (all related to promises). Tell me if you'd prefer them to be split in separate PRs.

Note: I amended my first commit to get a clean history.

tbroyer commented on src/Core/Commander/Providers/PanoramicProvider.js:
```diff
@@ -72,24 +70,18 @@
-
-        var that = this;
-        if(!_panoramicsMetaData){
+    PanoramicProvider.prototype.getMetaDataFromPos = function(longitude, latitude/*, distance*/){
```
Wasn't sure what to do here: entirely remove the unused `distance` parameter or just comment it out?

FWIW, it was passed to `getClosestPanoInMemory` which didn't _expect_ it; and I inlined `getClosestPanoInMemory` here in the promise's `then`.

nosy-b replied:
Yes it was originally set to ensure a max distance. Doesn't matter anymore

tbroyer replied:
OK, removed. (and directly amended the commit, given the nature and scope of the change)

tbroyer committed:
Fix one concurrency bug in PanoramicProvider wrt resource loading

Ensure that the resource is only loaded once over the network (even
when getMetaDatFromPos is called more than once before the request
returns).

This does NOT fix another concurrency bug where the PanoramicProvider's
state is modified from a promise callback with a value depending on
method argument values; as this requires a redesign of the API.

peppsac commented:
r+, merging.

Thanks !
